internal/rbd/replicationcontrollerserver.go

```diff
@@ -0,0 +1,349 @@
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		err = rbdVol.enableImageMirroring(ctx, mirroringMode)
+		if err != nil {
+			util.ErrorLog(ctx, "failed to enable mirroring of %s with error %s", err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+	return &replication.EnableVolumeReplicationResponse{}, nil
+}
+
+// DisableVolumeReplication disables replication on volume.
+func (cs *ControllerServer) DisableVolumeReplication(ctx context.Context,
+	req *replication.DisableVolumeReplicationRequest,
+) (*replication.DisableVolumeReplicationResponse, error) {
+	var rbdVol = &rbdVolume{}
+	var err error
+	// ger rbdVolume from VolumeID
+	rbdVol, err = cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol)
+	if err != nil {
+		return nil, err
+	}
+
+	// extract the force option
+	force, err := getForceOption(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		err = fmt.Errorf("failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State == librbd.MirrorImageEnabled {
+		err = rbdVol.disableImageMirroring(ctx, force)
+		if err != nil {
+			util.ErrorLog(ctx, "failed to enable mirroring of %s with error %s", err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		// TODO check do we need to return abort after disabling the mirroring
+		return &replication.DisableVolumeReplicationResponse{}, nil
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageDisabled {
+		return nil, status.Errorf(codes.Aborted, "Image is in %d Mode", mirroringInfo.State)
+	}
+
+	return &replication.DisableVolumeReplicationResponse{}, nil
+}
+
+// PromoteVolume promotes the volume as primary.
+func (cs *ControllerServer) PromoteVolume(ctx context.Context,
+	req *replication.PromoteVolumeRequest,
+) (*replication.PromoteVolumeResponse, error) {
+	var rbdVol = &rbdVolume{}
+	var err error
+	// ger rbdVolume from VolumeID
+	rbdVol, err = cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol)
+	if err != nil {
+		return nil, err
+	}
+
+	// extract the force option
+	force, err := getForceOption(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		err = fmt.Errorf("failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		return nil, status.Errorf(codes.InvalidArgument, "image is in %d Mode", mirroringInfo.State)
+	}
+
+	// promote secondary to primary or force promote image if its already
+	// primary
+	if !mirroringInfo.Primary || force && mirroringInfo.Primary {
```

Comment:
I thought we need to force promote in rbd only if the image was not already primary, IOW if if is alreayd primary then there is no promotion to be done. (which basically changes the condition here, but depends on the correctness of my understanding)

Reply:
Yes, you are correct. even with force, the promotion fails if the image is already primary. removed the check now.